packages/node-runtime-worker-thread/src/index.ts

```diff
@@ -0,0 +1,119 @@
+import { ChildProcess } from 'child_process';
+import { MongoClientOptions } from '@mongosh/service-provider-core';
+import { Runtime } from '@mongosh/browser-runtime-core';
+import { EvaluationListener } from '@mongosh/shell-evaluator';
+import spawnChildFromSource from './spawn-child-from-source';
+import { Caller, createCaller, exposeAll, WithClose } from './rpc';
+import type { WorkerRuntime as WorkerThreadWorkerRuntime } from './worker-runtime';
+import childProcessProxySrc from 'inline-entry-loader!./child-process-proxy';
+
+type ChildProcessRuntime = Caller<WorkerThreadWorkerRuntime>;
+
+class WorkerEvaluationListener {
+  exposedListener: WithClose<EvaluationListener>;
+
+  constructor(workerRuntime: WorkerRuntime, childProcess: ChildProcess) {
+    this.exposedListener = exposeAll<EvaluationListener>(
+      {
+        onPrompt(question, type) {
+          if (workerRuntime.evaluationListener?.onPrompt) {
+            return workerRuntime.evaluationListener.onPrompt(question, type);
+          }
+
+          return '';
+        },
+        onPrint(value) {
+          if (workerRuntime.evaluationListener?.onPrint) {
+            workerRuntime.evaluationListener.onPrint(value);
+          }
+        },
+        toggleTelemetry(enabled) {
+          if (workerRuntime.evaluationListener?.toggleTelemetry) {
+            workerRuntime.evaluationListener.toggleTelemetry(enabled);
+          }
+        },
+        onClearCommand() {
+          if (workerRuntime.evaluationListener?.onClearCommand) {
+            workerRuntime.evaluationListener.onClearCommand();
+          }
+        },
+        onExit() {
+          if (workerRuntime.evaluationListener?.onExit) {
+            return workerRuntime.evaluationListener.onExit();
+          }
+
+          return Promise.resolve() as Promise<never>;
+        }
+      },
+      childProcess
+    );
+  }
+}
+
+class WorkerRuntime implements Runtime {
+  private initOptions: {
+    uri: string;
+    driverOptions: MongoClientOptions;
+    cliOptions: { nodb?: boolean };
+  };
+
+  evaluationListener: EvaluationListener | null = null;
+
+  private childProcessEvaluationListener!: WorkerEvaluationListener;
+
+  private childProcess!: ChildProcess;
+
+  private childProcessRuntime!: ChildProcessRuntime;
+
+  private initWorkerPromise: Promise<void>;
+
+  constructor(
+    uri: string,
+    driverOptions: MongoClientOptions = {},
+    cliOptions: { nodb?: boolean } = {}
+  ) {
+    this.initOptions = { uri, driverOptions, cliOptions };
+    this.initWorkerPromise = this.initWorker();
+  }
+
+  private async initWorker() {
```

Comment:
Btw, I’m a bit surprised that the linter is okay with leaving out return value descriptions… I thought it complained about that for other packages

Reply:
Hmmm, I can't see the rule enabled in our eslint config, and it's not a recommended one 🤔 Should we add it?

Reply:
No strong feelings about that, but I personally like being explicit about the signature :woman_shrugging:

Reply:
Yeah, I like the rule too and (as it's pretty obvious from this PR, haha) always forget to add explicit return types when the rule is not there 😄